sys/include/firmware.h

```diff
@@ -0,0 +1,288 @@
+
+/**
+ *  @brief Length of secret Ed25519 key
+ */
+#define FIRMWARE_SECKEY_LEN         (crypto_sign_SECRETKEYBYTES)
+
+/**
+ *  @brief Length of public Ed25519 key
+ */
+#define FIRMWARE_PUBKEY_LEN         (crypto_sign_PUBLICKEYBYTES)
+
+/**
+ *  @brief Number of metadata bytes that get checksummed
+ */
+#define FIRMWARE_CHECKSUM_LEN       (20)
+
+/**
+ *  @brief Length of bytes to be signed
+ */
+#define FIRMWARE_SIGN_BYTES         (FIRMWARE_CHECKSUM_LEN + 4 + SHA256_DIGEST_LENGTH)
+
+/**
+ *  @brief Length of metadata prefixed to firmware binaries
+ */
+#ifndef FIRMWARE_METADATA_SIZE
+#define FIRMWARE_METADATA_SIZE      (256)
+#endif
+
+/**
+ *  @brief Length of padding to align metadata size
+ */
+#define FIRMWARE_PADDING            (FIRMWARE_METADATA_SIZE-(FIRMWARE_SIGN_BYTES + FIRMWARE_SIG_LEN))
+
+/**
+ *  @brief Real length of metadata without padding
+ */
+#define FIRMWARE_METADATA_REALLEN   (sizeof(firmware_metadata_t) - FIRMWARE_PADDING)
+
+/**
+ * @brief Structure to store firmware metadata
+ * @{
+ */
+typedef struct {
+    uint32_t magic_number;              /**< metadata magic_number (always "RIOT")  */
+    uint32_t appid;                     /**< Integer representing the application ID*/
+    uint32_t version;                   /**< Integer representing firmware version  */
+    uint32_t size;                      /**< Size of firmware image (w/o metadata)  */
+    uint32_t start_addr;                /**< Start address in flash                 */
+    uint32_t chksum;                    /**< checksum of metadata                   */
+    uint8_t hash[SHA256_DIGEST_LENGTH]; /**< SHA256 Hash of firmware image          */
+    uint8_t sig[FIRMWARE_SIG_LEN];      /**< Firmware signature                     */
+    uint8_t pad[FIRMWARE_PADDING];      /**< padding to total of
+                                             FIRMWARE_METADATA_SIZE bytes           */
+} firmware_metadata_t;
+/** @} */
+
+/**
+ * @brief  Print formatted FW image metadata to STDIO
+ *
+ * @param[in] metadata  ptr to firmware metadata
+ *
+ */
+void firmware_metadata_print(firmware_metadata_t *metadata);
+
+/**
+ * @brief  Validate FW image metadata
+ *
+ * @param[in] metadata  ptr to firmware metadata
+ *
+ */
+int firmware_validate_metadata_checksum(firmware_metadata_t *metadata);
+
+/**
+ * @brief  Calculate metadata checksum
+ *
+ * @param[in] metadata  ptr to firmware metadata
+ *
+ */
+uint32_t firmware_metadata_checksum(firmware_metadata_t *metadata);
+
+/**
+ * @brief  Sign metadata
+ *
+ * @param[in] metadata  ptr to firmware metadata
+ * @param[in] sk        NaCL secret signing key to use
+ *
+ */
+int firmware_sign_metadata(firmware_metadata_t *metadata, unsigned char *sk);
+
+/**
+ * @brief  Validate FW metadata signature
+ *
+ * @param[in] metadata  ptr to firmware metadata
+ * @param[in] pk        NaCL public signing key to use
+ *
+ */
+int firmware_validate_metadata_signature(firmware_metadata_t *metadata, const unsigned char *pk);
```

Comment:
I wanted to know what the return value means but unfortunately it's not documented.